contributors/design-proposals/storage/container-storage-interface.md

```diff
@@ -0,0 +1,460 @@
+
+To deploy a containerized third-party CSI volume driver, it is recommended that storage vendors:
+
+  * Create a “CSI volume driver” container that implements the volume plugin behavior and exposes a gRPC interface via a unix domain socket, as defined in the CSI spec (including Controller, Node, and Identity services).
+  * To deploy a CSI plugin, a cluster admin should deploy the following
+  * The Kubernetes team will provide helper containers (external-attacher, external-provisioner, Kubernetes CSI Helper) which will assist the “CSI volume driver” container in interacting with the Kubernetes system.
+    * StatefulSet with replica size 1, that should
+      * A StatefulSet (unlike a ReplicaSet) will guarantee that no more than 1 instance of the pod will be running at once (so we don’t have to worry about multiple instances of the external-provisioner or external-attacher in the cluster).
+      * Contain the following containers
+        * The “CSI volume driver” container created by the storage vendor.
+        * The external-attacher container provided by the Kubernetes team.
+        * The external-provisioner container provided by the Kubernetes team.
+      * Have the following volumes:
+        * `emptyDir` volume
+          * Mount inside all containers at `/var/lib/csi/sockets/pluginproxy/`
+          * The “CSI volume driver” container should create its Unix Domain Socket in this directory to enable communication with the Kubernetes helper container(s) (external-provisioner, external-attacher).
+    * DaemonSet
+      * A DaemonSet will ensure that 1 instance of the pod is deployed on every node (to facilitate communication from every instance of kubelet).
+      * Contain the following containers
+        * The “CSI volume driver” container created by the storage vendor.
+        * The “Kubernetes CSI Helper” container provided by the Kubernetes team responsible for registering the unix domain socket with kubelet and NodeId initialization.
+      * Have the following volumes:
+        * `hostpath` volume
+          * Expose `/var/lib/kubelet/device-plugins/kubelet.sock` from the host.
+          * Mount only in “Kubernetes CSI Helper” container at `/var/lib/csi/sockets/kubelet.sock`
+          * The Kubernetes to CSI proxy container will use this unix domain socket to register the CSI driver’s unix domain socket with kubelet.
+        * `hostpath` volume
+          * Expose `/var/lib/kubelet/` from the host.
+          * Mount only in “CSI volume driver” container at `/var/lib/kubelet/`
+          * Ensure [bi-directional mount propagation](https://kubernetes.io/docs/concepts/storage/volumes/#mount-propagation) is enabled, so that any mounts setup inside this container are propagated back to the host machine.
+        * `hostpath` volume
+          * Expose `/var/lib/kubelet/plugins/csi/sockets/kubeletproxy.sock` from the host.
```

Comment:
This approach does not specify how the k8s kubelet will handle multiple CSI Drivers running each at its own UDS address.  It seems that each CSI Driver should have its own UDS path indicated in the path like `/var/lib/kubelet/plugins/csi/<DriverName>/kubeletproxy.sock`.

Reply:
Description of this socket is weird. It shouldn't be exposed "from the host" but "to the host". It's the socket that CSI driver creates and listens on. We have very little control about this socket. It can have any name (as long as it's available to kubelet) and it's registered in Kubernetes by "Kubernetes CSI Helper" using `/var/lib/kubelet/device-plugins/kubelet.sock` (described few lines above).

I expect that CSI driver will have cmdline parameter (env. variable, config file option, ...) where it should create this socket. "Kubernetes CSI Helper" will have similar cmdline option (env. variable, ...) that says where expect CSI driver's socket. And author of the DaemonSet puts it together and makes sure that all the containers use the same socket path. It must be configurable and it must be different for each plugin.

Reply:
@jsafrane thanks with the flow you outline.  Completely agree with your assessment of about the way socket name will be registered and announced to the in-tree CSIDriver.  What about simply add the socket location in the PV Spec ?

Reply:
> What about simply add the socket location in the PV Spec 

Different nodes may have different socket paths for the same CSI driver, I'd like to keep only the driver name in PV spec and each kubelet can do its own mapping to the socket path.

Reply:
To make this workable until the socket registration part is done, I added a line saying that for alpha the volume driver must drop it's socket to `/var/lib/kubelet/plugins/csi/sockets/[driverName]/kubeletproxy.sock`. And kubelet can deterministically generate and examine this path for a given VolumeSource. 

For beta, we can do the whole socket registration.